juju/juju
# 2.2 bootstrap to subnet 1659640

## Description of change

This implements support for "subnet=" as a placement directive on AWS. This allows you to do "juju bootstrap --to subnet=A.B.C.D/24" and have Juju match that to a specific subnet to provision a machine in. It works for bootstrap, and I believe it will also work for any instance that you provision thereafter. Since Subnet in AWS also constrains you to an Availability Zone it also has that effect.

Note, this doesn't have Unit tests yet, as it was mostly meant as an exploration as to what was possible, and it turned out to be quite easy to just implement. I'd like to add tests before we would actually land this.

## QA steps

Create a bunch of subnets in your AWS VPC. Bootstrap with
```
  juju bootstrap aws/REGION --config vpc-id=VPC --to subnet=SUBNET
```

You can use --debug to see some of the details about the selection process. Or just note in your EC2 dashboard what subnet the resulting instance ends up in.

SUBNET can be a Name, subnet-id, or CIDR.

I haven't tested exactly what happens with "juju deploy --to" or "juju add-unit --to", though both should use the same logic (and it essentially just piggybacks on all our existing code to handle AvailabilityZones.)

## Documentation changes

This would affect what a user can specify during "juju bootstrap", and if we have explicit documentation about using Spaces in AWS, we should definitely include this information.

## Bug reference

[lp:#1659640](https://bugs.launchpad.net/juju/+bug/1659640)
